OsmAnd/src/net/osmand/plus/routing/SuggestionsMapsProvider.java

```diff
@@ -0,0 +1,142 @@
+package net.osmand.plus.routing;
+
+import net.osmand.IndexConstants;
+import net.osmand.Location;
+import net.osmand.data.LatLon;
+import net.osmand.map.WorldRegion;
+import net.osmand.plus.OsmandApplication;
+import net.osmand.plus.onlinerouting.EngineParameter;
+import net.osmand.plus.onlinerouting.OnlineRoutingHelper;
+import net.osmand.plus.onlinerouting.engine.EngineType;
+import net.osmand.plus.onlinerouting.engine.OnlineRoutingEngine;
+import net.osmand.util.Algorithms;
+
+import org.json.JSONException;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.LinkedHashSet;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+
+import static net.osmand.util.MapUtils.movePointToDistance;
+
+public class SuggestionsMapsProvider {
+
+	private static final int DISTANCE = 20000;
+	RouteCalculationParams params;
+	private boolean pointOnWater;
+
+	public SuggestionsMapsProvider(RouteCalculationParams params) {
+		this.params = params;
+	}
+
+	public boolean isPointOnWater() {
+		return pointOnWater;
+	}
+
+	public boolean checkIfObjectDownloaded(String downloadName) {
+		final String regionName = Algorithms.capitalizeFirstLetterAndLowercase(downloadName)
+				+ IndexConstants.BINARY_MAP_INDEX_EXT;
+		final String roadsRegionName = Algorithms.capitalizeFirstLetterAndLowercase(downloadName) + ".road"
+				+ IndexConstants.BINARY_MAP_INDEX_EXT;
+		return params.ctx.getResourceManager().getIndexFileNames().containsKey(regionName) || params.ctx.getResourceManager().getIndexFileNames().containsKey(roadsRegionName);
+	}
+
+	public List<Location> getLocationBasedOnDistanceInterval(List<Location> points) {
+		List<Location> mapsBasedOnPoints = new ArrayList<>();
+		for (int i = 0; i < points.size(); i++) {
+			int nextIndex = i + 1 < points.size() ? i + 1 : i;
+			mapsBasedOnPoints.add(0, points.get(i));
+			mapsBasedOnPoints.add(mapsBasedOnPoints.size(), points.get(nextIndex));
+			while (mapsBasedOnPoints.get(0).distanceTo(mapsBasedOnPoints.get(mapsBasedOnPoints.size() - 1)) > DISTANCE) {
+				float bearing = mapsBasedOnPoints.get(0).bearingTo(mapsBasedOnPoints.get(mapsBasedOnPoints.size() - 1));
```

Comment:
too many repetitions mapsBasedOnPoints.get(0) and mapsBasedOnPoints.get(mapsBasedOnPoints.size() - 1)